### ISSUE TITLE
Compilation error under c++20 with constexpr

### ISSUE THREAD

**Issue by maksymilian-palka** [boostorg/static_string#30]
Hello,

I have compilation errors for the following code:
```
#include <iostream>
#include <boost/static_string.hpp>

constexpr auto getString()
{
    boost::static_strings::static_string<1024> content;
    content.append("Hello world");
    return content;
}

constexpr auto msg = getString();

int main()
{
    std::cout << msg.c_str() << std::endl;
    return 0;
}
```

The code fails to compile under:
* clang 13.1.6
* VS2022 17.2.6

The walk around I've found is to allow an array initialisation of `static_string_base` class, meaning replacing:
```
#ifdef BOOST_STATIC_STRING_CPP20
    value_type data_[N + 1];
#else
    value_type data_[N + 1] {};
#endif
```
with
```
    value_type data_[N + 1] {};
```

**Comment by beached** [comment]
It looks like the issue is that boost config is using BOOST_STATIC_STRING_STANDARD_VERSION which is essentially __cplusplus with fixes for MSVC.  This does not tell us that the compiler supports anything in C++20 other than the -std=c++20 flag has been set.  I think for the init stuff in C++20, there is a feature flag set by the compiler `__cpp_constexpr_dynamic_alloc >= 201907L`, but not 100% that is the correct flag.

**Comment by beached** [comment]
Ah ok, I see now.  I thought it might have came with the allocator support in [More Constexpr Containers](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0784r7.html), I didn't know about [Permitting trivial default initialization in constexpr contexts](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1331r2.pdf)

**Comment by alandefreitas** [comment]
This should have been fixed in #32
